pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,488 @@
+			klog.V(3).Infof("Found internal IP for machine %q: %q", machine.GetName(), machineInternalAddress)
+			break
+		}
+	}
+
+	if machineInternalAddress == "" {
+		klog.Warningf("not found internal IP for machine %q", machine.GetName())
+		return nil, nil
+	}
+
+	nodes, err := r.listNodesByFieldFunc(nodeInternalIPIndex, machineInternalAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+
+	if len(nodes) > 0 {
+		klog.V(3).Infof("Found node %q for machine %q with internal IP %q", nodes[0].GetName(), machine.GetName(), machineInternalAddress)
+		return nodes[0].DeepCopy(), nil
+	}
+
+	klog.V(3).Infof("not found node match for machine %q with internal IP %q", machine.GetName(), machineInternalAddress)
+	return nil, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNode(node *corev1.Node) (*mapiv1beta1.Machine, error) {
+	if node == nil {
+		klog.V(3).Infof("Finding machine from node nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding machine from node %q", node.GetName())
+	machine, err := r.findMachineFromNodeByProviderID(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find machine from node %q by ProviderID: %v", node.GetName(), err)
+	}
+	if machine != nil {
+		return machine, nil
+	}
+
+	machine, err = r.findMachineFromNodeByIP(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by internal IP: %v", machine.GetName(), err)
+	}
+	return machine, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNodeByProviderID(node *corev1.Node) (*mapiv1beta1.Machine, error) {
+	if node == nil {
+		klog.V(3).Infof("Finding machine from node nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding machine from node %q by ProviderID", node.GetName())
+	if node.Spec.ProviderID == "" {
+		klog.Warningf("Node %q has no providerID", node.GetName())
+		return nil, nil
+	}
+
+	machines, err := r.listMachinesByFieldFunc(machineProviderIDIndex, node.Spec.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+	if len(machines) > 0 {
+		klog.V(3).Infof("Found machine %q for node %q with providerID %q", machines[0].GetName(), node.GetName(), node.Spec.ProviderID)
+		return machines[0].DeepCopy(), nil
+	}
+	return nil, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNodeByIP(node *corev1.Node) (*mapiv1beta1.Machine, error) {
+	if node == nil {
+		klog.V(3).Infof("Finding machine from node nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding machine from node %q by IP", node.GetName())
+	var nodeInternalAddress string
+	for _, a := range node.Status.Addresses {
+		if a.Type == corev1.NodeInternalIP {
+			nodeInternalAddress = a.Address
+			klog.V(3).Infof("Found internal IP for node %q: %q", node.GetName(), nodeInternalAddress)
+			break
+		}
+	}
+
+	if nodeInternalAddress == "" {
+		klog.Warningf("not found internal IP for node %q", node.GetName())
```

Comment:
s/not/Not

Reply:
Should we error here?

Reply:
nit: "Node %q has no internal IP address"

Reply:
> Should we error here?

Isn't the state transient here? Next time it may have addresses.